tests/test_precompiled.py

```diff
@@ -438,6 +439,29 @@ def test_precompiled_sanitise_pdf_without_notify_tag(client, auth_header):
     )
 
 
+def test_precompiled_sanitise_pdf_for_pdf_with_content_near_margins(client, auth_header):
```

Comment:
This test may not flag a bunch of issues:

- We [perform validation / count pages](https://github.com/alphagov/notifications-template-preview/blob/03df0d92522f13091b081f3fe04c188e85d2ade6/app/precompiled.py#L174-L179) before rewriting, so if the rewrite code trashes the PDF we may not know, as the error could still be the same.

- Some of [the rewriting code](https://github.com/alphagov/notifications-template-preview/blob/03df0d92522f13091b081f3fe04c188e85d2ade6/app/precompiled.py#L224-L228) won't run on this sample file, as it doesn't need CMYK or unembedded font conversion (I manually checked the predicates).

So I'm not sure we get much from this test. I was thinking about how to address the overall concern of "something might trash the PDF and we wouldn't know". We could use a different technique: output fixtures. This would involve passing a PDF - maybe the example DWP one - through the code and saving the result, so that we can check the bytes are identical each time. Candidate functions:

- `rewrite_address_block`
- `convert_pdf_to_cmyk`
- `embed_fonts`
- `add_notify_tag_to_letter`

As long as we make it clear how to regenerate the output fixtures (and do a visual check they look OK), then these tests could provide a lot more reassurance beyond the text-snippet-oriented ones we have at the moment.

What do you think?